Translate Forms (#61)
reactjs/de.reactjs.org: content/docs/forms.md

  - "docs/forms-zh-CN.html"
---

> HTML-Formularelemente funktionieren in React ein bisschen anders als DOM-Elemente, weil Formularelemente einen natürlichen internen State besitzt. Beispielsweise akzeptiert dieses HTML-Formular einen einzelnen Namen:

Reviewer: ```suggestion
HTML-Formularelemente funktionieren in React ein bisschen anders als DOM-Elemente, weil Formularelemente einen natürlichen internen State besitzen. Beispielsweise akzeptiert dieses HTML-Formular einen einzelnen Namen:
```

```html
<form>
  <label>
    Name:
    <input type="text" name="name" />
  </label>
  <input type="submit" value="Absenden" />
</form>
```

Dieses Formular hat das Standard HTML Verhalten, dass es beim Absenden den Benutzer zu einer neuen Seite weiterleitet. Wenn du dieses Verhalten in React willst, funktioniert es einfach. In den meisten Fällen ist es jedoch sinnvoll, eine JavaScript-Funktion zu benutzen, die das Absenden des Formulars übernimmt und Zugriff auf die Eingabedaten des Benutzers hat. Um dies zu erreichen, wird standardmäßig eine Technik namens "kontrollierte Komponenten" verwendet.

## Kontrollierte Komponenten {#controlled-components}

Formelemente wie `<input>`, `<textarea>` und `<select>` behalten in HTML typischerweise ihren eigenen Zustand bei und aktualisiert ihn basierend auf Benutzereingaben. In React wird der veränderbare State typischerweise in der State-Eigenschaft der Komponente gehalten und nur mit [`setState()`](/docs/react-component.html#setstate) aktualisiert.

> Wir können beides kombinieren und den React-State zur alleinigen "source of truth" machen. Die React-Komponente rendert dann das Formular und steuert ebenso die nachfolgenden Benutzereingaben. Ein Eingabeformular-Element, dessen Wert auf diese Art und Weise von React kontrolliert wird, wird "kontrollierte Komponente" genannt.

Reviewer: ```suggestion
Wir können beides kombinieren, indem wir de React-State zur alleinigen "source of truth" machen. Die React-Komponente rendert dann das Formular und steuert ebenso die nachfolgenden Benutzereingaben. Ein Eingabeformular-Element, dessen Wert auf diese Art und Weise von React kontrolliert wird, wird "kontrollierte Komponente" genannt.
```

          Name:
          <input type="text" value={this.state.value} onChange={this.handleChange} />
        </label>
        <input type="submit" value="Absenden" />
      </form>
    );
  }
}
```

[**Auf CodePen ausprobieren**](https://codepen.io/gaearon/pen/VmmPgp?editors=0010)

> Da das `value`-Attribut unserer Formular-Element gesetzt wurde, entspricht der angezeigte Wert immer `this.state.value` und macht somit den React-State zur "source of truth". Da `handleChange` bei jedem Tastendruck ausgeführt wird, um den React-State zu aktualisieren, wird der Wert es ebenso, sobald Benutzereingaben stattfinden.

Reviewer: ```suggestion
Da das `value`-Attribut unseres Formular-Elementes gesetzt wurde, entspricht der angezeigte Wert immer `this.state.value` und macht somit den React-State zur "source of truth". Da `handleChange` bei jedem Tastendruck ausgeführt wird, um den React-State zu aktualisieren, wird der Wert es ebenso, sobald Benutzereingaben stattfinden.
```

> Bei einer kontrollierten Komponente hat jede Änderung des States eine dazugehörige Handler-Funktion. Dadurch ist es einfacher Benutzereingaben zu ändern oder zu validieren. Wenn wir zum Beispiel immer wollen, dass die Namen in GRoßbuchstaben geschrieben werden, können wir `handleChange` so schreiben:

Reviewer: ```suggestion
Bei einer kontrollierten Komponente hat jede Änderung des States eine dazugehörige Handler-Funktion. Dadurch ist es einfacher Benutzereingaben zu ändern oder zu validieren. Wenn wir zum Beispiel immer wollen, dass die Namen in Großbuchstaben geschrieben werden, können wir `handleChange` so schreiben:
```

```javascript{2}
handleChange(event) {
  this.setState({value: event.target.value.toUpperCase()});
}
```

## Der textarea-Tag {#the-textarea-tag}

In HTML definiert ein `<textarea>`-Element seinen Text durch seine Kinder:

```html
<textarea>
  Hallo, dies ist ein bisschen Text im textarea-Tag
</textarea>
```

In React benutzt eine `<textarea>` stattdessen das `value`-Attribut. Auf diese Weise kann ein Formular, das eine `<textarea>` benutzt, sehr ähnlich geschrieben werden wie ein Formular, das ein einziges Eingabefeld verwendet:

```javascript{4-6,12-14,26}
class EssayForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {

>       value: 'Bitte schreibe einen Aufsatz über dein lieblings DOM-Element.'

Reviewer: ```suggestion
      value: 'Bitte schreibe einen Aufsatz über dein Lieblings-DOM-Element.'
```

Reviewer: https://www.duden.de/suchen/dudenonline/lieblings

  }

  handleSubmit(event) {
    alert('Ein Aufsatz wurde eingereicht: ' + this.state.value);
    event.preventDefault();
  }

  render() {
    return (
      <form onSubmit={this.handleSubmit}>
        <label>
          Aufsatz:
          <textarea value={this.state.value} onChange={this.handleChange} />
        </label>
        <input type="submit" value="Absenden" />
      </form>
    );
  }
}
```

Merke dir, dass `this.state.value` im Konstruktor initialisert wird, so dass das Textfeld mit etwas Text gefüllt wird.

## Der select-Tag {#the-select-tag}

> In HTML, erstellt `<select>` eine Dropdown-Liste. Zum Beispiel erstellt dieses HTML eine Dropdown-Liste von Geschmacksrichtungen:

Reviewer: ```suggestion
In HTML erstellt `<select>` eine Dropdown-Liste. Zum Beispiel erstellt dieses HTML eine Dropdown-Liste von Geschmacksrichtungen:
```

```html
<select>
  <option value="grapefruit">Pampelmuse</option>
  <option value="lime">Limette</option>
  <option selected value="coconut">Kokosnuss</option>
  <option value="mango">Mango</option>
</select>
```

> Beachte, dass die Option Kokosnuss aufgrund des `selected`-Attributs zunächst ausgewählt ist. React benutzt statt dem `selected`-Attribut, ein `value`-Attribut auf dem `select`-Tag. Dies ist in einer kontrollierten Komponente komfortabler, da du sie nur an einer Stelle bearbeiten musst. Zum Beispiel:

Reviewer: ```suggestion
Beachte, dass die Option "Kokosnuss" aufgrund des `selected`-Attributs zunächst ausgewählt ist. React benutzt statt dem `selected`-Attribut, ein `value`-Attribut auf dem `select`-Tag. Dies ist in einer kontrollierten Komponente komfortabler, da du sie nur an einer Stelle bearbeiten musst. Zum Beispiel:
```

        <label>
          Pick your favorite flavor:
          <select value={this.state.value} onChange={this.handleChange}>
            <option value="grapefruit">Pampelmuse</option>
            <option value="lime">Limette</option>

>             <option selected value="coconut">Kokosnuss</option>

Reviewer: ```suggestion
            <option value="coconut">Kokosnuss</option>
```

Reviewer: No need for `selected`.

            <option value="mango">Mango</option>
          </select>
        </label>
        <input type="submit" value="Absenden" />
      </form>
    );
  }
}
```

[**Auf CodePen ausprobieren**](https://codepen.io/gaearon/pen/JbbEzX?editors=0010)

> Zusammenfassend ist es so, dass `<input type="text">`, `<textarea>`, und `<select>` sehr ähnlich funkionieren - alle besitzen ein `value`-Attribut, welches benutzt werden um eine kontrollierte Komponente umzusetzen.

Reviewer: ```suggestion
Zusammenfassend ist es so, dass `<input type="text">`, `<textarea>` und `<select>` sehr ähnlich funktionieren - alle besitzen ein `value`-Attribut, welches benutzt wird, um eine kontrollierte Komponente umzusetzen.
```

> Hinweis
>
> Du kannst ein Array an das `value`-Attribut übergeben um somit mehrere Optionen im `select`-Tag auszuwählen:
>
>```js
><select multiple={true} value={['B', 'C']}>
>```

## Der file input-Tag {#the-file-input-tag}

> In HTML, an `<input type="file">` lets the user choose one or more files from their device storage to be uploaded to a server or manipulated by JavaScript via the [File API](https://developer.mozilla.org/en-US/docs/Web/API/File/Using_files_from_web_applications).

Reviewer: Sollte noch übersetzt werden, danke.

```html
<input type="file" />
```

Da dessen Wert schreibgeschützt ist, ist dies eine **unkontrollierte** Komponente in React. Es wird später zusammen mit anderen unkontrollierten Komponenten [in der Dokumentation](/docs/uncontrolled-components.html#the-file-input-tag) behandelt.

## Umgang mit mehreren Eingabefeldern {#handling-multiple-inputs}

> Wenn du mehrere kontrollierte `input`-Elemente benötigst, kannst du das `name`-Attribut zu jedem Element hinzufügen und die Handler-Funktion entscheiden lassen was baiserend auf dem Wert von `event.target.name`, zu tun ist.

Reviewer: ### 
```suggestion
Wenn du mehrere kontrollierte `input`-Elemente benötigst, kannst du das `name`-Attribut zu jedem Element hinzufügen und die Handler-Funktion entscheiden lassen, was, basierend auf dem Wert von `event.target.name`, zu tun ist.
```

    return (
      <form>
        <label>

>           Wirf hingehen:

Reviewer: ```suggestion
          Wird hingehen:
```

}
```

[**Auf CodePen ausprobieren**](https://codepen.io/gaearon/pen/wgedvV?editors=0010)

> Beachte, dass wir die ES6-Syntax für [berechnete Bezeichnernamen](https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Operators/Objekt_Initialisierer#Berechnete_Bezeichnernamen) verwenden, um den State ensprechend dem gegebenem Namen zu ändern:

Reviewer: `computed property name` eventuell auf Englisch belassen?

```js{2}
this.setState({
  [name]: value
});
```

> Ist das Equivalent zu diesem ES5-Code:

Reviewer: ```suggestion
Ist das Äquivalent zu diesem ES5-Code:
```

```js{2}
var partialState = {};
partialState[name] = value;
this.setState(partialState);
```

Da `setState()` automatisch [die einzelne Teile in den aktuellen State überführt](/docs/state-and-lifecycle.html#state-updates-are-merged), benötigen wir nur den Aufruf der geänderten Teile:

## Gesteuerte Null-Wert-Eingaben {#controlled-input-null-value}

Das Angeben eines `value`-Prop auf einer [kontrollierten Komponente(/docs/forms.html#controlled-components) verhindert, dass der Benutzer den Eingabewert ändern kann, es sei denn, du wünscht es. Wenn du einen Wert angegeben hast, aber das Eingabefeld noch editierbar ist, hast du möglicherweise versehentlich `value` auf `undefiniert` oder `null` gesetzt.

> Der folgende Code demonstriert dies. (Das Eingabefeld ist zuerste gesperrt, wird aber nach einer kurzen Verzögerung editierbar.)

Reviewer: ```suggestion
Der folgende Code demonstriert dies. (Das Eingabefeld ist zuerst gesperrt, wird aber nach einer kurzen Verzögerung editierbar.)
```

```

## Alternativen zu kontrollierten Komponenten {#alternatives-to-controlled-components}

Es kann manchmal mühsam sein kontrollierte Komponenten zu verwenden, da du für jede Art und Weise der Änderung deiner Daten einen Event-Handler bereitstellen und den gesamten Eingabe-State durch die React-Komponente leiten musst. Dies kann besonders ärgerlich sein, wenn du eine bestehende Codebasis zu React konvertieren möchtest oder eine Nicht-React-Bibliothek in einer React-Anwendung integrierst. In diesen Situationen solltest du vielleicht [unkontrollierte Komponenten](/docs/uncontrolled-components.html) ausprobieren, eine alternative Technik zur Implementierung von Eingabeformularen.

## Vollumfassende Lösungen {#fully-fledged-solutions}

> Wenn du nach einer Komplettlösung mit Validierung, dem im Blick behalten von besuchten Feldern und der Handhabung von Formularverabeitung suchst, ist [Formik](https://jaredpalmer.com/formik) eine der beliebtesten Entscheidungen. Es basiert jedoch auf den selben Pinzipien von kontrollierten Komponenten und der Verwaltung des States - also vernachlässige es nicht, darüber etwas zu lernen.

Reviewer: ```suggestion
Wenn du nach einer Komplettlösung mit Validierung, dem Im-Blick-Behalten von besuchten Feldern und der Handhabung von Formularverarbeitung suchst, ist [Formik](https://jaredpalmer.com/formik) eine der beliebtesten Entscheidungen. Es basiert jedoch auf den selben Pinzipien von kontrollierten Komponenten und der Verwaltung des States - also vernachlässige es nicht, darüber etwas zu lernen.
```

Reviewer: https://www.korrekturen.de/sprachleben/rechtschreibung/der_horror_mit_den_substantivischen_zusammensetzungen.shtml

Author: :D ich les es mir durch
